ksql-engine/src/main/java/io/confluent/ksql/util/TableFunctionExpressionRewriter.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.util;
+
+import io.confluent.ksql.engine.rewrite.ExpressionTreeRewriter.Context;
+import io.confluent.ksql.execution.expression.tree.ColumnReferenceExp;
+import io.confluent.ksql.execution.expression.tree.Expression;
+import io.confluent.ksql.execution.expression.tree.FunctionCall;
+import io.confluent.ksql.execution.expression.tree.VisitParentExpressionVisitor;
+import io.confluent.ksql.function.FunctionRegistry;
+import io.confluent.ksql.name.ColumnName;
+import io.confluent.ksql.schema.ksql.ColumnRef;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+public class TableFunctionExpressionRewriter
+    extends VisitParentExpressionVisitor<Optional<Expression>, Context<Void>> {
+
+  private int variableIndex = 0;
+  private final FunctionRegistry functionRegistry;
+
+  public TableFunctionExpressionRewriter(final FunctionRegistry functionRegistry) {
+    super(Optional.empty());
+    this.functionRegistry = functionRegistry;
+  }
+
+  @Override
+  public Optional<Expression> visitFunctionCall(
+      final FunctionCall node,
+      final Context<Void> context) {
+    final String functionName = node.getName().name();
+    if (functionRegistry.isTableFunction(functionName)) {
+      final ColumnName varName = ColumnName.udtfColumn(variableIndex);
+      variableIndex++;
+      return Optional.of(
+          new ColumnReferenceExp(node.getLocation(), ColumnRef.of(Optional.empty(), varName)));
```

Comment:
It's basically very similar to the AggregateExpressionRewriter.

Reply:
> If there's a criticism of this approach it should also apply to the existing aggregate function implementation

Don't worry, I think both @agavra and myself are aware this is based heavily on existing code that does similar things.  Following the patterns that are already there is the most obvious approach for implementing something like this. However, it's sometimes worth thinking if the pattern can be improved before replicating it.  There are a lot of areas in the KSQL code base that could do with an overhaul. Copying such less-than-perfect patterns and abstractions should be something we strive to avoid, well, at least most of the time! ;)

As for this particular issue...  not sure, it's been a while since my head has been in this code.  Though, I don't think UDFs do anything similar to this. UDTFs are closer in nature to UDFs than UDAFs.  So it's probably worth using UDTs for the template of implementing UDTFs, rather than the current UDAF template.   I think that may result in cleaner code.

Reply:
It was boring on the flight, so I spent some time thinking about this (and refactoring things).

In the previous code PlanNodes are allowed to change schema, but they aren't allowed to change select expressions. This led me to do that weird if check in LogicalPlanner to get the final select expressions, it also leads AggregateNode to do it's own final projection.

Some nodes need to change the select items - flatmap and aggregate are two examples so PlanNodes need someway of propagating the changed select items. I've therefore added the passing of select items into the constructors of PlanNodes and an ability to get the (potentially changed) select items for the next node in the change. This allows us to do the final projection for flatmap in a much cleaner way, and it will also allow us to refactor out AggregateNode into several other nodes and for it to reuse ProjectNode too. The refactor of AggregateNode can be done in another PR.